wdl-engine/src/backend/tes.rs

```diff
@@ -0,0 +1,537 @@
+        self.memory
+    }
+
+    async fn run(self, spawned: oneshot::Sender<()>) -> Result<TaskExecutionResult> {
+        // Create the attempt directory
+        let attempt_dir = self.inner.attempt_dir();
+        fs::create_dir_all(attempt_dir).with_context(|| {
+            format!(
+                "failed to create directory `{path}`",
+                path = attempt_dir.display()
+            )
+        })?;
+
+        // Write the evaluated command to disk
+        // This is done even for remote execution so that a copy exists locally
+        let command_path = attempt_dir.join(COMMAND_FILE_NAME);
+        fs::write(&command_path, self.inner.command()).with_context(|| {
+            format!(
+                "failed to write command contents to `{path}`",
+                path = command_path.display()
+            )
+        })?;
+
+        let task_dir = format!(
+            "{name}-{timestamp}/",
+            name = self.name,
+            timestamp = chrono::Utc::now().format("%Y%m%d-%H%M%S")
+        );
+
+        // Start with the command file as an input
+        let mut inputs = vec![
+            Input::builder()
+                .path(GUEST_COMMAND_PATH)
+                .contents(Contents::Path(command_path.to_path_buf()))
+                .ty(InputType::File)
+                .read_only(true)
+                .build(),
+        ];
+
+        for input in self.inner.inputs() {
+            // Currently, if the input is a directory with contents, we'll error evaluation
+            // TODO: in the future, we should be uploading the entire contents to cloud
+            // storage
+            if input.kind() == InputKind::Directory {
+                if let EvaluationPath::Local(path) = input.path() {
+                    if let Ok(mut entries) = path.read_dir() {
+                        if entries.next().is_some() {
+                            bail!(
+                                "cannot upload contents of directory `{path}`: operation is not \
+                                 yet supported",
+                                path = path.display()
+                            );
+                        }
+                    }
+                }
+                continue;
+            }
+
+            // TODO: for local files, upload to cloud storage rather than specifying the
+            // input contents directly
+            inputs.push(
+                Input::builder()
+                    .path(input.guest_path().expect("should have guest path"))
+                    .contents(match input.path() {
+                        EvaluationPath::Local(path) => Contents::Path(path.clone()),
+                        EvaluationPath::Remote(url) => Contents::Url(url.clone()),
+                    })
+                    .ty(input.kind())
+                    .read_only(true)
+                    .build(),
+            );
+        }
+
+        let outputs_url = self
+            .config
+            .outputs
```

Comment:
Does this mean that this is going to unwrap if the key is not provided in the configuration object? If so, do we want to treat it (and all similar scenarios) in a bit more user-friendly way?

Reply:
Currently the value is a validated as a required field of the backend's configuration and validation is performed upon the backend's construction.

Ideally it'll be a truly optional field in the future where if an output URL is not specified, we treat it as a shared file system path and treat them as local paths instead.

Reply:
Right, I now am coming back from later in my review where I see that. I think it deserves a `SAFETY` comment to indicate that to the reader.